File: models/ensemble.py
"""Ensemble: stack multiple models, return mean + uncertainty."""

import numpy as np

from models.base import OptionPricer


class EnsemblePricer(OptionPricer):
    """Ensemble pricer that combines multiple sub-pricers.

    Returns mean, std, and per-model outputs for uncertainty quantification.
    """

    def __init__(self, name: str = "ensemble", models=None):
        super().__init__(name)
        self.models = models or []

    def add_model(self, model: OptionPricer):
        self.models.append(model)

    @property
    def is_trained(self):
        return len(self.models) > 0

    def price(self, S, K, T, r=0.045, q=0):
        if not self.models:
            return 0.0
        prices = [m.price(S, K, T, r, q) for m in self.models if hasattr(m, 'is_trained') and m.is_trained]
        if not prices:
            return 0.0
        return float(np.mean(prices))

    def price_with_uncertainty(self, S, K, T, r=0.045, q=0):
        """Return (mean_price, std_price, per_model_prices)."""
        prices = []
        details = []
        for m in self.models:
            if hasattr(m, 'is_trained') and m.is_trained:
                p = m.price(S, K, T, r, q)
                prices.append(p)
                details.append({"model": m.name, "price": p})
        if not prices:
            return 0.0, 0.0, []
        return float(np.mean(prices)), float(np.std(prices)), details

    def surface(self, strikes, expiries, S, r=0.045, q=0):
        if not self.models:
            return np.zeros((len(expiries), len(strikes)), dtype=np.float32)
        surfaces = []
        for m in self.models:
            if hasattr(m, 'is_trained') and m.is_trained:
                surfaces.append(m.surface(strikes, expiries, S, r, q))
        if not surfaces:
            return np.zeros((len(expiries), len(strikes)), dtype=np.float32)
        return np.mean(surfaces, axis=0)

    def surface_with_uncertainty(self, strikes, expiries, S, r=0.045, q=0):
        """Return (mean_surface, std_surface)."""
        surfaces = []
        for m in self.models:
            if hasattr(m, 'is_trained') and m.is_trained:
                surfaces.append(m.surface(strikes, expiries, S, r, q))
        if not surfaces:
            NE, NS = len(expiries), len(strikes)
            return np.zeros((NE, NS)), np.zeros((NE, NS))
        return np.mean(surfaces, axis=0), np.std(surfaces, axis=0)

    def greeks(self, S, K, T, r=0.045, q=0):
        if not self.models:
            return {"delta": 0.0, "gamma": 0.0, "vega": 0.0, "theta": 0.0}
        greeks_list = [m.greeks(S, K, T, r, q) for m in self.models
                       if hasattr(m, 'is_trained') and m.is_trained]
        if not greeks_list:
            return {"delta": 0.0, "gamma": 0.0, "vega": 0.0, "theta": 0.0}
        return {k: float(np.mean([g[k] for g in greeks_list])) for k in greeks_list[0]}

    def save(self, path):
        import json
        meta = {
            "name": self.name,
            "model_names": [m.name for m in self.models],
        }
        with open(path, "w") as f:
            json.dump(meta, f)

    def load(self, path):
        pass  # Sub-models must be loaded individually

File: src/models/ensemble.js
// Ensemble: stack multiple models, return mean + uncertainty.
// Port of models/ensemble.py.

const { OptionPricer } = require('./base');
const { mean, std, zeros2d } = require('../util');

function trained(m) {
  return m && m.isTrained;
}

class EnsemblePricer extends OptionPricer {
  constructor(name = 'ensemble', models = null) {
    super(name);
    this.models = models || [];
  }

  addModel(model) {
    this.models.push(model);
  }

  get isTrained() {
    return this.models.length > 0;
  }

  price(S, K, T, r = 0.045, q = 0) {
    const prices = this.models.filter(trained).map((m) => m.price(S, K, T, r, q));
    if (!prices.length) return 0;
    return mean(prices);
  }

  priceWithUncertainty(S, K, T, r = 0.045, q = 0) {
    const prices = [];
    const details = [];
    for (const m of this.models) {
      if (trained(m)) {
        const p = m.price(S, K, T, r, q);
        prices.push(p);
        details.push({ model: m.name, price: p });
      }
    }
    if (!prices.length) return { mean: 0, std: 0, details: [] };
    return { mean: mean(prices), std: std(prices), details };
  }

  surface(strikes, expiries, S, r = 0.045, q = 0) {
    const surfaces = this.models.filter(trained).map((m) => m.surface(strikes, expiries, S, r, q));
    const NE = expiries.length;
    const NS = strikes.length;
    if (!surfaces.length) return zeros2d(NE, NS);
    return meanSurfaces(surfaces, NE, NS);
  }

  greeks(S, K, T, r = 0.045, q = 0) {
    const list = this.models.filter(trained).map((m) => m.greeks(S, K, T, r, q));
    if (!list.length) return { delta: 0, gamma: 0, vega: 0, theta: 0 };
    const keys = Object.keys(list[0]);
    const out = {};
    for (const key of keys) out[key] = mean(list.map((g) => g[key]));
    return out;
  }
}

function meanSurfaces(surfaces, NE, NS) {
  const out = zeros2d(NE, NS);
  for (let i = 0; i < NE; i++) {
    for (let j = 0; j < NS; j++) {
      let s = 0;
      for (const surf of surfaces) s += surf[i][j];
      out[i][j] = s / surfaces.length;
    }
  }
  return out;
}

module.exports = { EnsemblePricer };
